src/api/blocks.py

```diff
@@ -0,0 +1,390 @@
+        """
+        current_hash = hash(self._blockstate)
+
+        if self._extra_blocks:
+            current_hash = current_hash + hash(self._extra_blocks)
+
+        return current_hash
+
+    def __add__(self, other: Block) -> Block:
+        """
+        Allows for other Block objects to be added to this Block object's ``extra_blocks``
+
+        :param other: The Block object to add to the end of this Block object's `extra_blocks`
+        :return: A new Block object with the same data but with an additional Block at the end of ``extra_blocks``
+        """
+        if (
+            len(other._extra_blocks) == 0
+        ):  # Reduces the amount of extra objects/references created
+            other_cpy = other
+        else:
+            other_cpy = Block(other._namespace, other._base_name, other._properties)
+
+        other_extras = []
+        for eb in other._extra_blocks:
+            if (
+                len(eb.extra_blocks) == 0
+            ):  # Reduces the amount of extra objects/references created
+                other_extras.append(eb)
+            else:
+                other_extras.append(Block(eb._namespace, eb._base_name, eb._properties))
+
+        return Block(
+            self._namespace,
+            self._base_name,
+            self._properties,
+            [*self._extra_blocks, other_cpy, *other_extras],
+        )
+
+    def __sub__(self, other: Block) -> Block:
+        """
+        Allows for other Block objects to be subtracted from this Block object's ``extra_blocks``
+
+        :param other: The Block object to subtract from this Block objects' ``extra_blocks``
+        :return: A new Block object without any instances of the subtracted block in ``extra_blocks``
+        """
+        if (
+            len(other._extra_blocks) == 0
+        ):  # Reduces the amount of extra objects/references created
+            other_cpy = other
+        else:
+            other_cpy = Block(other._namespace, other._base_name, other._properties)
+
+        other_extras = []
+        for eb in other._extra_blocks:
+            if len(eb.extra_blocks) == 0:
+                other_extras.append(eb)
+            else:
+                other_extras.append(Block(eb._namespace, eb._base_name, eb._properties))
+
+        # Sets are unordered, so a regular set subtraction doesn't always return the order we want (it sometimes will!)
+        # So we loop through all of our extra blocks and only append those to the new_extras list if they aren't in
+        # extra_blocks_to_remove
+        new_extras = []
+        extra_blocks_to_remove = (other_cpy, *other_extras)
+        for eb in self._extra_blocks:
+            if eb not in extra_blocks_to_remove:
+                new_extras.append(eb)
+
+        return Block(self._namespace, self._base_name, self._properties, new_extras)
+
+    def __iadd__(self, other: Block) -> TypeError:
+        raise TypeError("You cannot add an extra block to an already existing block")
```

Comment:
If this is doing what I am thinking it does it forces the block class to be immutable

Reply:
The python documentation only specifies that __iadd__ "should attempt" do do the operation inplace. If it is not possible it can still return a new object. This is was the builtin `str` does. It's immutable, but `str1 += str2` still works. It's not modifying `str1`, but creating a new string and storing it back in `str1` (`str1 = str1 + str2`)

Leaving `__iadd__` and `__isub__` undefined makes it fall back to `__add__`/`__sub__`, which still keeps them immutable, but allows the use of the `+=`/`-=` operators